### PR TITLE
Allow custom ports for services and checks when using driver address_mode

### LINKED ISSUE #3380: Allow health checks to use exposed container ports on driver-assigned ip.
It's great to see the new address_mode service parameter that allows a service to be registered with a driver-managed IP instead of the host IP and it would be super-helpful if this capability was extended to the health checks.

My understanding is that it is actually consul that performs the health checks and my Consul agents have access to the same internal network Docker uses to assign IPs to containers. It would be great if the health checks could therefore use this internal IP - the same one used to register the service with Consul when the address_mode is set to 'driver'.

Similarly, my docker containers expose ports but don't map ports to the host - they are on an internal network with a proxy (Fabio) handling access from the outside world. But this means there is no port stanza for most of my containers. In the case, as explained above, where a driver-allocated IP is used for a health-check, it would be extremely useful to also specify an actual port number rather than a port label so health checks could be performed on exposed ports.

Maybe the best approach would be to have some new check types such as container_tcp and container_http that behave in the way described.

### PR DESCRIPTION
**Pull request by schmichael**
Fixes #3380

Adds address_mode to checks (but no auto) and allows services and checks
to set literal port numbers when using address_mode=driver.

This allows SDNs, overlays, etc to advertise internal and host addresses
as well as do checks against either.

### `task_runner.go` changes

Fixes a bug found when using fewer mocks in TaskRunner tests: changing only an interpolated variable used in a service or a check wouldn't actually update the service or check.

### CODE REVIEW

**Comment by schmichael** [comment]
Some test binaries for the brave:

[linux_amd64.zip](https://github.com/hashicorp/nomad/files/1532666/linux_amd64.zip)
[windows_amd64.zip](https://github.com/hashicorp/nomad/files/1532668/windows_amd64.zip)

File: client/driver/mock_driver.go [hunk 1]
```diff
@@ -286,7 +326,12 @@ func (h *mockDriverHandle) run() {
 	for {
 		select {
 		case <-timer.C:
-			close(h.doneCh)
+			select {
+			case <-h.doneCh:
+				// already closed
+			default:
+				close(h.doneCh)
```
Comment (schmichael) [inline client/driver/mock_driver.go]: I ran into a double close panic on the mock driver when testing, so I guarded all of these closes. Not sure why we had never hit it before...

File: client/driver/mock_driver.go [hunk 2]
```diff
@@ -133,7 +162,8 @@ func (m *MockDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse
 	}
 	m.logger.Printf("[DEBUG] driver.mock: starting task %q", task.Name)
 	go h.run()
-	return &StartResponse{Handle: &h}, nil
+
+	return &StartResponse{Handle: &h, Network: net}, nil
```
Comment (schmichael) [inline client/driver/mock_driver.go]: `mock_driver` now supports DriverNetwork! I should have added this when I added DriverNetwork originally...

File: client/task_runner_test.go [hunk 3]
```diff
@@ -73,15 +73,17 @@ func (m *MockTaskStateUpdater) String() string {
...
-	tr       *TaskRunner
-	allocDir *allocdir.AllocDir
-	vault    *vaultclient.MockVaultClient
-	consul   *mockConsulServiceClient
+	upd          *MockTaskStateUpdater
+	tr           *TaskRunner
+	allocDir     *allocdir.AllocDir
+	vault        *vaultclient.MockVaultClient
+	consul       *consul.MockAgent
+	consulClient *consul.ServiceClient
```
Comment (schmichael) [inline client/task_runner_test.go]: Removing this use of `*mockConsulServiceClient` is what exposed the task runner bug.

There's still a few tests that use that mock as they inspect the order of Consul operations which the mock makes very easy.

File: command/agent/consul/client.go [hunk 4]
```diff
@@ -1079,3 +1083,44 @@ func isNomadService(id string) bool {
 	const prefix = nomadServicePrefix + "-executor"
 	return strings.HasPrefix(id, prefix)
 }
+
```
Comment (schmichael) [inline command/agent/consul/client.go]: Below is the core address resolution logic for both services and checks.

File: jobspec/parse_test.go [hunk 5]
```diff
@@ -580,6 +628,7 @@ func TestParse(t *testing.T) {
 			for _, d := range pretty.Diff(actual, tc.Result) {
 				t.Logf(d)
 			}
+			//t.Logf(pretty.Sprint(actual))
```
Comment (dadgar) [inline jobspec/parse_test.go]: Remove

File: website/source/docs/job-specification/service.html.md [hunk 6]
```diff
@@ -120,6 +122,11 @@ the script will run inside the Docker container. If your task is running in a
 chroot, it will run in the chroot. Please keep this in mind when authoring check
 scripts.
 
+- `address_mode` `(string: "host")` - Same as `address_mode` on `service`.
```
Comment (dadgar) [inline website/source/docs/job-specification/service.html.md]: I would add a full example for this and link it to this

File: nomad/structs/structs.go [hunk 7]
```diff
@@ -3444,7 +3462,12 @@ func validateServices(t *Task) error {
 		knownServices[service.Name+service.PortLabel] = struct{}{}
 
 		if service.PortLabel != "" {
-			servicePorts[service.PortLabel] = append(servicePorts[service.PortLabel], service.Name)
+			if _, err := strconv.Atoi(service.PortLabel); service.AddressMode == "driver" && err == nil {
```
Comment (preetapan) [inline nomad/structs/structs.go]: This is hard to read, and I am not sure if the logic is correct. If addressmode is not driver, and the strconv returns an error (if portlabel is *not* an integer), then should the error really be ignored? if so, why even do the parsing?

**Comment by thetooth** [comment]
I have setup a test nomad job but am not getting working health checks:
`agent: socket connection failed ':0': dial tcp :0: getsockopt: connection refused`
```
      service {
        name = "poc-server"
        address_mode = "driver"
        port = "6379"
        check {
          type     = "tcp"
          interval = "10s"
          timeout  = "2s"
        }
      }
```

File: website/source/docs/job-specification/service.html.md [hunk 8]
```diff
@@ -99,19 +99,21 @@ does not automatically enable service discovery.
...
   driver-specific) this service should advertise. `host` indicates the host IP
-  and port. `driver` advertises the IP used in the driver (e.g. Docker's internal
-  IP) and uses the ports specified in the port map. The default is `auto` which
-  behaves the same as `host` unless the driver determines its IP should be used.
-  This setting supported Docker since Nomad 0.6 and rkt since Nomad 0.7. It
-  will advertise the container IP if a network plugin is used (e.g. weave).
+  and port. `driver` advertises the IP used in the driver (e.g. Docker's
+  internal IP) and uses the container's port specified in the port map. The
+  default is `auto` which behaves the same as `host` unless the driver
+  determines its IP should be used.  This setting supported Docker since Nomad
```
Comment (preetapan) [inline website/source/docs/job-specification/service.html.md]: Supported _in_ Docker

File: website/source/docs/job-specification/service.html.md [hunk 9]
```diff
@@ -99,19 +99,21 @@ does not automatically enable service discovery.
...
+  0.6 and rkt since Nomad 0.7. It will advertise the container IP if a network
+  plugin is used (e.g. weave).
```
Comment (preetapan) [inline website/source/docs/job-specification/service.html.md]: I would make each valid option sub items or a table with option/behavior as headings, and 1/2 lines explaining them. I would also start with the default option auto.

File: command/agent/consul/client.go [hunk 10]
```diff
@@ -1079,3 +1083,44 @@ func isNomadService(id string) bool {
...
+		if port, ok := driverNet.PortMap[portLabel]; ok {
+			return driverNet.IP, port, nil
+		}
+
+		// If port isn't a label, try to parse it as a literal port number
+		port, err := strconv.Atoi(portLabel)
+		if err != nil {
+			return "", 0, fmt.Errorf("invalid port %q: %v", portLabel, err)
+		}
+
```
Comment (preetapan) [inline command/agent/consul/client.go]: Should this have a check for port > 0 - If some job file misconfiguration leads to someone specifying 0 as the port label, we should not try to use it and fail fast..

File: website/source/docs/job-specification/service.html.md (hunk 6 above)
Comment (schmichael) [reply website/source/docs/job-specification/service.html.md]: Added a full section at the bottom of that page with a link

**Comment by schmichael** [comment]
@thetooth Thanks for testing! Fixed in f4e341e

Since `checks` inherit the service's port (a numeric port in this case) but do *not* inherit the `address_mode` - the numeric port `6379` was being looked as a port label due to the check's default `host` address mode. I didn't have validation to catch this error and the empty value (`0`) was being used for the port!

I can see that it's a bit confusing to have `port` inherited by `check` but not `address_mode`, but I think the behavior matches many people's configuration (Consul is running on the host so service checks should be against the host address). Let me know if there's anything other than this validation that could help prevent such errors!

Here's another build:

[linux_amd64.zip](https://github.com/hashicorp/nomad/files/1541620/linux_amd64.zip)

**Comment by thetooth** [comment]
@schmichael Some usability issues:

The following configuration does not work as expected. Since I am not using a port map I tried a deploy without a port label in resources,
```
      service {
        name = "poc-server"
        address_mode = "driver"
        port = "6379"
        check {
          type     = "tcp"
          interval = "10s"
          timeout  = "2s"
        }
      }
      resources {
        cpu    = 512 # MHz 
        memory = 512 # MB 
        network {
          mbits = 10
        }
      }
```
The error for this was `unable to get address for check "service: \"poc-server\" check": invalid port "6379": port label not found`, so then I defined a port label:
```
          port "redis" {
            static = 6379
          }
```
Which cased `unable to get address for service "poc-server": invalid port "redis": strconv.Atoi: parsing "redis": invalid syntax`, finally I changed the port label to the port itself:
```
          port "6379" {
            static = 6379
          }
```
This worked and I have working service checks to the docker network(weave mesh) however as I hinted at in my earlier comment, since the port label is used docker has been instructed to publish those ports to localhost: `127.0.0.1:6379->6379/tcp, 127.0.0.1:6379->6379/udp`. It would be very useful from a security perspective to not have services exposed to local sessions as an artifact of using health checks.

File: website/source/docs/job-specification/service.html.md [hunk 11]
```diff
@@ -99,19 +99,22 @@ does not automatically enable service discovery.
...
-  behaves the same as `host` unless the driver determines its IP should be used.
-  This setting supported Docker since Nomad 0.6 and rkt since Nomad 0.7. It
-  will advertise the container IP if a network plugin is used (e.g. weave).
+  and port. `driver` advertises the IP used in the driver (e.g. Docker's
+  internal IP) and uses the container's port specified in the port map. The
+  default is `auto` which behaves the same as `host` unless the driver
+  determines its IP should be used.  This setting is supported in Docker since
+  Nomad 0.6 and rkt since Nomad 0.7. Nomad will advertise the container IP if a
+  network plugin is used (e.g. weave). See [below for
+  details.](#using-driver-address-mode)
```
Comment (preetapan) [inline website/source/docs/job-specification/service.html.md]: Did you see the other comment I had about making each of the modes sub list headers for clarity?

File: command/agent/consul/client.go [hunk 12]
```diff
@@ -1079,3 +1090,50 @@ func isNomadService(id string) bool {
...
+		if port, ok := driverNet.PortMap[portLabel]; ok {
+			return driverNet.IP, port, nil
+		}
+
+		// If port isn't a label, try to parse it as a literal port number
+		port, err := strconv.Atoi(portLabel)
+		if err != nil {
+			return "", 0, fmt.Errorf("invalid port %q: %v", portLabel, err)
+		}
+		if port == 0 {
```
Comment (preetapan) [inline command/agent/consul/client.go]: Change to <=0

**Comment by schmichael** [comment]
@thetooth Thanks for testing. The `check` is inheriting `port=6379` from the `service` but not the service's address mode. This is by design as we want to maintain the old behavior of `check.address_mode=host` by default.

However, the error you're receiving is entirely unhelpful! The fix is to specify `address_mode = "driver"` on the check or expose a port via a resource & port map on the host for the `check` to use. I'll try to update the validation to reflect this.

**Comment by schmichael** [comment]
Not sure why `TestConfig_Listener` is failing on Travis, but it passes for me locally. Merging this :grimacing:  

```
--- FAIL: TestConfig_Listener (0.02s)
	config_test.go:524: err: listen tcp 127.0.0.1:24000: bind: address already in use
```

**Pull request merged**